### PR TITLE
Adjust to a DI structure

### PR DESCRIPTION
**Pull request by acupofjose**
Ref to #36 and supabase-community/supabase-csharp#23

- Creates 3 interfaces: `IGotrueApi`, `IGotrueClient` and `IGotrueStatelessClient`
- Keeps a large amount of backwards compatibility with previous API
- Can be instantiated using `Microsoft.Extensions.Hosting.Host`

Breaking Changes:
- `StatelessClient` is no longer a `static` class but can be used as a `transient` class (allows for maintaining api via Interface)
- `Client` requires calling `InstantiateAsync` but **is no longer a singleton**.
- Several constants that were part of the `Client` class were moved to `Constants` for better organization.

Big thanks to [HunteRoi](https://github.com/HunteRoi) for his suggestions!

### CODE REVIEW

**Comment by acupofjose** [comment]
@HunteRoi could you look at this?

File: Gotrue/Client.cs [hunk 1]
```diff
@@ -160,29 +72,7 @@ public static Client Instance
 
         private Api api;
```
Comment (HunteRoi) [inline Gotrue/Client.cs:73]: IMO this should be an `IApi` instead of a concrete class

File: Gotrue/Client.cs [hunk 2]
```diff
@@ -194,27 +84,25 @@ public static void Initialize(ClientOptions options = null, Action<Client> callb
...
 
             // Retrieve the session
-            if (instance.ShouldPersistSession)
-                await instance.RetrieveSessionAsync();
+            if (ShouldPersistSession)
+                await RetrieveSessionAsync();
 
-            return instance;
+            return this;
         }
```
Comment (HunteRoi) [inline Gotrue/Client.cs:106]: Could probably be moved to a constructor.
eg:
```csharp
public Client (IGotrueApi<User, Session> api, ClientOptions options) 
{
            Options = options;
            AutoRefreshToken = options.AutoRefreshToken;
            ShouldPersistSession = options.PersistSession;
            SessionPersistor = options.SessionPersistor;
            SessionRetriever = options.SessionRetriever;
            SessionDestroyer = options.SessionDestroyer;

            this.api = api;

            // Retrieve the session
            if (ShouldPersistSession)
                await RetrieveSessionAsync();
}

public async Task<IGotrueClient<User, Session>> InitializeAsync (ClientOptions options = null) 
{
            if (options == null)
                options = new ClientOptions();

            return new Client(new Api(options.Url, options.Headers), options);
}
```

**Review by HunteRoi — CHANGES_REQUESTED** [review]

File: Gotrue/Client.cs (hunk 2 above)
Comment (acupofjose) [reply Gotrue/Client.cs:106]: Hm. I've also wanted to move this to a constructor, but it needs that async call to `RetrieveSessionAsync()` - though I suppose we can separate that out of the constructor as a change.

File: Gotrue/Client.cs (hunk 1 above)
Comment (acupofjose) [reply Gotrue/Client.cs:73]: Agreed!

File: Gotrue/Client.cs (hunk 2 above)
Comment (HunteRoi) [reply Gotrue/Client.cs:106]: oops did not see that await! You're right though

**Comment by acupofjose** [comment]
@HunteRoi if this looks good to you - this will be the structure I'll be taking to update the other libraries. (Now tracking at supabase-csharp#35)

**Comment by HunteRoi** [comment]
Went through it diagonally and it seems good

**Pull request merged**